x/spec/types/spec.go

```diff
@@ -7,6 +7,97 @@ import (
 
 const minCU = 1
 
+type getterFunc func(index string) (Spec, bool)
+ 
+// ExpandSpec expands the "imports" field of the spec -if needed, recursively-
+// and adds to the current Spec those additional APIs from the imported Spec(s).
+// It takes as argument a getter function to fetch imported Spec(s) by name from
+// the keeper store, on demand. (Can't import the keeper module due to circular
+// dependencies)
+func (spec *Spec) ExpandSpec(getSpec getterFunc) (map[string]string, error) {
+	depends := map[string]bool{spec.Index: true}
+	details := map[string]string{"imports": fmt.Sprintf("%s", spec.Index)}
+	err := spec.doExpandSpec(getSpec, depends, details)
+	return details, err
+}
+
+// doExpandSpec performs the actual work and recusion for ExpandSpec above.
+func (spec *Spec) doExpandSpec(getSpec getterFunc, depends map[string]bool, details map[string]string) error {
+	if len(spec.Imports) == 0 {
+		return nil
+	}
+
+	var parents []Spec
+
+	// visual markers when import deepens
+	details["imports"] += "->["
+
+	// recursion to get all parent specs
+	comma := ""
+	for _, index := range spec.Imports {
+		imported, found := getSpec(index)
+		// import of unknown Spec not allwoed
+		if !found {
+			details["imports"] += fmt.Sprintf("%s%s(unknown)", comma, index)
+			return fmt.Errorf("imported spec unknown: %s", index)
+		}
+		// import of disabled Spec not allowed
+		if !imported.Enabled {
+			details["imports"] += fmt.Sprintf("%s%s(disabled)", comma, index)
+			return fmt.Errorf("imported spec disabled: %s", index)
+		}
+
+		details["imports"] += fmt.Sprintf("%s%s", comma, index)
+
+		// loop in the recursion not allowed
+		if _, found := depends[index]; found {
+			return fmt.Errorf("import loops not allowed for spec: %s", index)
+		}
+
+		depends[index] = true
+		if err := imported.doExpandSpec(getSpec, depends, details); err != nil {
+			return err
+		}
+		delete(depends, index)
+
+		parents = append(parents, imported)
+		comma = ","
+	}
+
+	details["imports"] += "]"
+
+	var mergedApis []ServiceApi
+	mergedApisMap := make(map[string]ServiceApi)
+
+	// collect all parents' Specs' APIs
+	for _, imported := range parents {
+		for _, api := range imported.Apis {
+			if api.Enabled {
+				// duplicate API(s) not allowed
+				if _, found := mergedApisMap[api.Name]; found {
+					return fmt.Errorf("duplicate imported api: %s (in spec: %s)", api.Name, imported.Index)
+				}
+				mergedApisMap[api.Name] = api
+				mergedApis = append(mergedApis, api)
+			}
+		}
+	}
+
+	currentApis := make(map[string]bool)
+	for _, api := range spec.Apis {
+		currentApis[api.Name] = true
+	}
+
+	// merge collected APIs into current spec's APIs (unless overridden)
+	for _, api := range mergedApis {
```

Comment:
we are missing the api deletion if it is disabled.
if the api is disabled dont add it to the list

Reply:
If an API is disabled in the parent, then it won't be added to the `mergedApis` list - see L75 above.